Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp

```diff
@@ -67,42 +66,29 @@ bool PersistentStore::write_data(int &pos, const uint8_t *value, size_t size, ui
     firstWrite = false;
   }
 
-  // First write full words
-  int i = 0;
-  int wordsToWrite = size / sizeof(uint16_t);
-  uint16_t* wordBuffer = (uint16_t *)value;
-  while (wordsToWrite) {
-    status = FLASH_ProgramHalfWord(pageBase + pos + (i * 2), wordBuffer[i]);
-    if (status != FLASH_COMPLETE) return true;
-    wordsToWrite--;
-    i++;
-  }
-
-  // Now, write any remaining single byte
-  const uint16_t odd = size & 1;
-  if (odd) {
-    uint16_t temp = value[size - 1];
-    status = FLASH_ProgramHalfWord(pageBase + pos + i, temp);
-    if (status != FLASH_COMPLETE) return true;
+  for (int i=0; i<size; i++) {
+    if (FLASH_ProgramHalfWord(EEPROM_PAGE0_BASE + (pos + i) * 2, value[i]) != FLASH_COMPLETE)
+      return true;
   }
 
   crc16(crc, value, size);
-  pos += size + odd;
+  pos += size;
   return false;
 }
 
 bool PersistentStore::read_data(int &pos, uint8_t* value, const size_t size, uint16_t *crc, const bool writing/*=true*/) {
-  for (uint16_t i = 0; i < size; i++) {
-    byte* accessPoint = (byte*)(pageBase + pos + i);
-    uint8_t c = *accessPoint;
-    if (writing) value[i] = c;
-    crc16(crc, &c, 1);
+  for (int i=0; i<size; i++) {
+    uint8_t v = *(uint16_t *)(EEPROM_PAGE0_BASE + (pos + i) * 2);
```

Comment:
this code line seems very strange to me... you put 2 bytes of data in only one byte...

Reply:
MSB is stripped out with implicit uint16 => uint8 casting;

Reply:
indeed
```
0007f0c0  ff ff ff ff ff ff ff ff  56 00 36 00 38 00 00 00  |........V.6.8...|
0007f0d0  ac 00 89 00 01 00 c8 00  00 00 00 00 00 00 c8 00  |................|
0007f0e0  00 00 00 00 00 00 64 00  00 00 00 00 00 00 b8 00  |......d.........|
0007f0f0  0b 00 00 00 00 00 20 00  4e 00 00 00 00 00 00 00  |...... .N.......|
0007f100  00 00 a0 00 42 00 00 00  00 00 a0 00 42 00 00 00  |....B.......B...|
0007f110  00 00 c8 00 43 00 00 00  00 00 c4 00 42 00 00 00  |....C.......B...|
0007f120  00 00 48 00 43 00 00 00  00 00 48 00 43 00 00 00  |..H.C.....H.C...|
0007f130  00 00 c8 00 42 00 00 00  00 00 c8 00 41 00 00 00  |....B.......A...|
0007f140  00 00 48 00 43 00 00 00  00 00 fa 00 43 00 00 00  |..H.C.......C...|
0007f150  00 00 48 00 43 00 00 00  00 00 00 00 00 00 00 00  |..H.C...........|
0007f160  00 00 00 00 00 00 00 00  00 00 20 00 41 00 00 00  |.......... .A...|
0007f170  00 00 20 00 41 00 cd 00  cc 00 cc 00 3e 00 00 00  |.. .A.......>...|
0007f180  00 00 a0 00 40 00 0a 00  d7 00 a3 00 3c 00 00 00  |....@.......<...|
0007f190  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1a0  00 00 00 00 00 00 01 00  00 00 00 00 00 00 00 00  |................|
0007f1b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1c0  03 00 03 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f1d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f200  00 00 00 00 00 00 00 00  00 00 00 00 8f 00 c2 00  |................|
0007f210  75 00 bf 00 00 00 00 00  00 00 00 00 00 00 00 00  |u...............|
0007f220  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f250  00 00 00 00 00 00 00 00  00 00 00 00 04 00 04 00  |................|
0007f260  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f280  00 00 00 00 c0 00 7f 00  00 00 00 00 c0 00 7f 00  |................|
*
0007f300  00 00 ff 00 0a 00 00 00  5a 00 00 00 00 00 c8 00  |........Z.......|
0007f310  00 00 fa 00 00 00 3c 00  00 00 64 00 00 00 00 00  |......<...d.....|
0007f320  00 00 8f 00 c2 00 89 00  41 00 0a 00 d7 00 23 00  |........A.....#.|
0007f330  3e 00 01 00 f0 00 e7 00  43 00 00 00 00 00 80 00  |>.......C.......|
0007f340  3f 00 14 00 00 00 e1 00  3a 00 a9 00 43 00 d3 00  |?.......:...C...|
0007f350  bc 00 23 00 41 00 00 00  ea 00 2e 00 45 00 20 00  |..#.A.......E. .|
0007f360  00 00 01 00 00 00 00 00  40 00 40 00 00 00 00 00  |........@.@.....|
0007f370  34 00 42 00 00 00 00 00  00 00 00 00 00 00 00 00  |4.B.............|
0007f380  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f390  50 00 41 00 00 00 00 00  00 00 00 00 00 00 00 00  |P.A.............|
0007f3a0  00 00 41 00 00 00 00 00  00 00 00 00 e0 00 3f 00  |..A...........?.|
0007f3b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f3e0  00 00 00 00 64 00 00 00  00 00 00 00 64 00 00 00  |....d.......d...|
0007f3f0  00 00 00 00 03 00 00 00  00 00 00 00 64 00 00 00  |............d...|
0007f400  00 00 00 00 64 00 00 00  00 00 00 00 03 00 00 00  |....d...........|
0007f410  00 00 00 00 03 00 00 00  00 00 00 00 1e 00 00 00  |................|
0007f420  00 00 00 00 1e 00 00 00  00 00 00 00 1e 00 00 00  |................|
*
0007f440  00 00 00 00 1e 00 00 00  00 00 00 00 00 00 00 00  |................|
0007f450  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f580  00 00 00 00 00 00 c8 00  42 00 00 00 00 00 00 00  |........B.......|
0007f590  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0007f5c0  40 00 40 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |@.@.............|
0007f5d0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00080000

```